### ISSUE TITLE
invalid mass fraction error when running flame wave model with aprox13 or triple alpha network.

### ISSUE THREAD

**Issue by zhichen3** [AMReX-Astro/Castro#2096]
I'm encountering the error: `amrex::Error::591::Invalid mass fraction in Castro::normalize_species() !!!`, when running the flame_wave model with aprox13 or triple_alpha_plus_cago network. This error occurred when I added `integrator.jacobian = 2` and `castro.retry_small_density_cutoff = 1` to the default input file in git, specifically: `inputs_He/inputs.He.1000Hz`. Whether or not these extra parameters caused the issue remains unknown.

**Comment by zingale** [comment]
update: this was not seen on GPUs / summit, because that check is disabled on GPUs.

Also, it seems that the problem is happening during the reflux step.

**Comment by zingale** [comment]
this is happening at the `expand_state()` call in `initialize_do_advance` midway through the level 1 advance (where level 1 is 4x finer than level 0)

**Comment by zingale** [comment]
even if we run with `castro.state_interp_order = 0`, we still hit this issue:

```
   
... compute diffusive term at level 1
   
... compute diffusive term at level 1
(i, j, k) = 2 24 0 , X[0] = 1.930137441
(i, j, k) = 2 24 0 , X[1] = -0.5402948753
(i, j, k) = 2 24 0 , X[12] = -0.3887283302
Castro::do_new_sources() time = 0.300138036
```

**Comment by zingale** [comment]
here's the backtrace:

```
 0: ./Castro2d.gnu.TEST.MPI.ex() [0x66b6b6]
    amrex::BLBackTrace::print_backtrace_info(_IO_FILE*)
/home/zingale/development/amrex//Src/Base/AMReX_BLBackTrace.cpp:179:39

 1: ./Castro2d.gnu.TEST.MPI.ex() [0x66d5b4]
    amrex::BLBackTrace::handler(int)
/home/zingale/development/amrex//Src/Base/AMReX_BLBackTrace.cpp:85:15

 2: ./Castro2d.gnu.TEST.MPI.ex() [0x43ee68]
    amrex::Error(char const*) inlined at ../../../Source/driver/Castro.cpp:3022:21 in Castro::normalize_species(amrex::MultiFab&, int
)
/home/zingale/development/amrex//Src/Base/AMReX.H:120:5
Castro::normalize_species(amrex::MultiFab&, int)
../../../Source/driver/Castro.cpp:3022:21

 3: ./Castro2d.gnu.TEST.MPI.ex() [0x44f80c]
    Castro::clean_state(amrex::MultiFab&, double, int) inlined at ../../../Source/driver/Castro.cpp:2851:42 in Castro::reflux(int, in
t)
../../../Source/driver/Castro.cpp:4329:16
Castro::reflux(int, int)
../../../Source/driver/Castro.cpp:2851:42

 4: ./Castro2d.gnu.TEST.MPI.ex() [0x4501db]
    Castro::post_timestep(int)
../../../Source/driver/Castro.cpp:1903:9

 5: ./Castro2d.gnu.TEST.MPI.ex() [0x6e3180]
    amrex::Amr::timeStep(int, double, int, int, double)
/home/zingale/development/amrex//Src/Amr/AMReX_Amr.cpp:2076:32

 6: ./Castro2d.gnu.TEST.MPI.ex() [0x6d9e2e]
    amrex::Amr::coarseTimeStep(double)
/home/zingale/development/amrex//Src/Amr/AMReX_Amr.cpp:2127:13

 7: ./Castro2d.gnu.TEST.MPI.ex() [0x428edd]
    main
../../../Source/driver/main.cpp:136:33

 8: /lib64/libc.so.6(+0x2d560) [0x7fb6d5fd5560]
    __libc_start_call_main
../sysdeps/nptl/libc_start_call_main.h:58:16

 9: /lib64/libc.so.6(__libc_start_main+0x7c) [0x7fb6d5fd560c]
    call_init inlined at ../csu/libc-start.c:379:5 in __libc_start_main@@GLIBC_2.34
../csu/libc-start.c:128:20
__libc_start_main_impl
../csu/libc-start.c:379:5

```

**Comment by zingale** [comment]
it is clearer what's happening if we do: `castro.update_sources_after_reflux = 0`, then there is no `clean_state` in the actual reflux.

If I explicitly call `normalize_species` before the reflux, then things are clean, but we crash at the `clean_state` after the reflux on Level 0 (presumably after accumulating all of the level 1 fluxes).

**Comment by zingale** [comment]
and indeed, if I set `castro.do_reflux=0` there are no problems.

So somehow the reflux is messing with the species here.

**Comment by zhichen3** [comment]
Update: recent changes in the development branch seem to solve the problem and I'm able to go over ~11,000 steps for aprox13 network, hence closing the issue.

**Comment by zhichen3** [comment]
It crashed again after ~14,000 steps unfortunately.

**Comment by zhichen3** [comment]
This issue should be solved with pull request #2207.
